File: port-forwarding.js
const net = require('net');

// Configuration
const LOCAL_HOST = '192.168.1.11';
const LOCAL_PORT = 3000;
const TARGET_HOST = '192.168.1.12';
const TARGET_PORT = 3000;

// Create the forwarding server
const server = net.createServer((clientSocket) => {
  // Connect to target server
  const targetSocket = net.connect({
    host: TARGET_HOST,
    port: TARGET_PORT
  }, () => {
    console.log(`Connected to target ${TARGET_HOST}:${TARGET_PORT}`);
    
    // Set up two-way data transfer
    clientSocket.pipe(targetSocket);
    targetSocket.pipe(clientSocket);
    
    // Handle client socket events
    clientSocket.on('error', (err) => {
      console.error('Client socket error:', err);
      targetSocket.end();
    });
    
    clientSocket.on('end', () => {
      console.log('Client disconnected');
      targetSocket.end();
    });
    
    // Handle target socket events
    targetSocket.on('error', (err) => {
      console.error('Target socket error:', err);
      clientSocket.end();
    });
    
    targetSocket.on('end', () => {
      console.log('Target disconnected');
      clientSocket.end();
    });
  });
  
  targetSocket.on('error', (err) => {
    console.error(`Error connecting to target ${TARGET_HOST}:${TARGET_PORT}:`, err);
    clientSocket.end();
  });
});

// Handle server errors
server.on('error', (err) => {
  console.error('Server error:', err);
});

// Start the server
server.listen(LOCAL_PORT, LOCAL_HOST, () => {
  console.log(`Port forwarding server running at ${LOCAL_HOST}:${LOCAL_PORT}`);
  console.log(`Forwarding connections to ${TARGET_HOST}:${TARGET_PORT}`);
});

// Handle script termination
process.on('SIGINT', () => {
  console.log('\nShutting down port forwarding server...');
  server.close();
  process.exit(0);
});

File: port-forwarding.py
import socket
import sys
import threading
import time

def forward_socket(source, target):
    """
    Forward data between source and target sockets.
    """
    while True:
        try:
            data = source.recv(4096)
            if not data:
                break
            target.send(data)
        except:
            break
    
    try:
        source.close()
    except:
        pass
    
    try:
        target.close()
    except:
        pass

def handle_client(client_socket):
    """
    Handle each client connection by forwarding to target host.
    """
    target_host = "192.168.1.12"
    target_port = 3000
    
    try:
        # Create connection to target
        target_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        target_socket.connect((target_host, target_port))
        
        # Create threads to forward data in both directions
        client_to_target = threading.Thread(
            target=forward_socket,
            args=(client_socket, target_socket)
        )
        
        target_to_client = threading.Thread(
            target=forward_socket,
            args=(target_socket, client_socket)
        )
        
        client_to_target.daemon = True
        target_to_client.daemon = True
        
        client_to_target.start()
        target_to_client.start()
        
        # Wait for one of the threads to finish
        while client_to_target.is_alive() and target_to_client.is_alive():
            time.sleep(0.5)
            
    except Exception as e:
        print(f"Error forwarding connection: {e}")
        try:
            client_socket.close()
        except:
            pass

def main():
    local_host = "192.168.1.11"  # Listen on this IP
    local_port = 3000  # Listen on this port
    
    server = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    
    try:
        # Reuse the socket if it's already in use
        server.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        server.bind((local_host, local_port))
        server.listen(5)
        
        print(f"[*] Listening on {local_host}:{local_port}")
        print(f"[*] Forwarding to 192.168.1.12:{local_port}")
        
        while True:
            client_socket, addr = server.accept()
            print(f"[*] Accepted connection from {addr[0]}:{addr[1]}")
            
            client_handler = threading.Thread(
                target=handle_client,
                args=(client_socket,)
            )
            client_handler.daemon = True
            client_handler.start()
            
    except KeyboardInterrupt:
        print("\n[*] Shutting down...")
        server.close()
        sys.exit(0)
    except Exception as e:
        print(f"[!] Error: {e}")
        server.close()
        sys.exit(1)

if __name__ == "__main__":
    main()